fv/netsets_test.go

```diff
@@ -0,0 +1,359 @@
+			srcNS := api.NewGlobalNetworkSet()
+			srcNS.Name = "ns-1"
+			srcNS.Spec.Nets = []string{
+				"10.65.0.0/24",
+				"10.65.1.0/24",
+				"10.65.2.0/24",
+			}
+			srcNS.Labels = map[string]string{
+				"allow-as-source": "",
+			}
+			client.GlobalNetworkSets().Create(utils.Ctx, srcNS, utils.NoOptions)
+
+			destNS := api.NewGlobalNetworkSet()
+			destNS.Name = "ns-2"
+			destNS.Spec.Nets = []string{
+				"10.65.2.0/24",
+				"10.65.3.0/24",
+			}
+			destNS.Labels = map[string]string{
+				"allow-as-dest": "",
+			}
+			client.GlobalNetworkSets().Create(utils.Ctx, destNS, utils.NoOptions)
+
+			cc.ExpectNone(w[0], w[1]) // not in dest net set
+			cc.ExpectSome(w[0], w[2])
+			cc.ExpectSome(w[0], w[3])
+			cc.ExpectNone(w[1], w[0]) // not in dest net set
+			cc.ExpectSome(w[1], w[2])
+			cc.ExpectSome(w[1], w[3])
+			cc.ExpectNone(w[2], w[0]) // not in dest net set
+			cc.ExpectNone(w[2], w[1]) // not in dest net set
+			cc.ExpectSome(w[2], w[3])
+			// 3 isn't in the source net set so it can't talk to anyone.
+			cc.ExpectNone(w[3], w[0])
+			cc.ExpectNone(w[3], w[1])
+			cc.ExpectNone(w[3], w[2])
+		})
+
+		AfterEach(func() {
+			for ii := range w {
+				w[ii].Stop()
+			}
+		})
+	})
+
+	Describe("churn tests", func() {
+		var (
+			policies    []*api.GlobalNetworkPolicy
+			networkSets []*api.GlobalNetworkSet
+		)
+
+		BeforeEach(func() {
+			// Start a workload so we have something to add policy to
+			w := workload.Run(
+				felix,
+				"w",
+				"cali12345",
+				"10.65.0.2",
+				"8055",
+				"tcp",
+			)
+			w.Configure(client)
+
+			// Generate policies and network sets.
+			for i := 0; i < numPolicies; i++ {
+				pol := api.NewGlobalNetworkPolicy()
+				pol.Name = fmt.Sprintf("policy-%d", i)
+				pol.Spec = api.GlobalNetworkPolicySpec{
+					Selector: "all()",
+					Ingress: []api.Rule{
+						{
+							Action: "Allow",
+							Source: api.EntityRule{
+								Selector: fmt.Sprintf("label == '%d'", i),
+							}},
+					},
+				}
+				policies = append(policies, pol)
+			}
+			for i := 0; i < numNetworkSets; i++ {
+				ns := api.NewGlobalNetworkSet()
+				ns.Name = fmt.Sprintf("netset-%d", i)
+				ns.Labels = map[string]string{
+					"foo":   "bar",
+					"baz":   "biff",
+					"label": fmt.Sprintf("%d", i%numPolicies),
+				}
+				nets := generateNets(i * 100)
+				ns.Spec = api.GlobalNetworkSetSpec{
+					Nets: nets,
+				}
+				networkSets = append(networkSets, ns)
+			}
+			log.SetLevel(log.InfoLevel)
+		})
+
+		churnPolicies := func(iterations int) {
+			log.Info("Churning policies...")
+			created := false
+			for i := 0; i < iterations; i++ {
+				if created {
+					for _, pol := range policies {
+						log.WithField("policy", pol.Name).Info("Deleting policy")
+						cxt, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+						_, err := client.GlobalNetworkPolicies().Delete(cxt, pol.Name, options.DeleteOptions{})
+						cancel()
```

Comment:
What does the cancel do here?  If the Delete call returns immediately, then (a) I don't see the point of the 10s timeout, and (b) we could be cancelling before the deletion has had any time to happen.

Reply:
The delete is synchronous but you're always supposed to call `cancel()` to correctly clean up the context.  In this case, I expect that calling cancel frees the resources associated with the timeout timer.

Reply:
OK, although that strikes me as quite weird.

In my experience, at least with k8s, the delete-that-matters (i.e. the resource actually disappearing) is asynchronous w.r.t. the Delete call (which returns immediately) and relatively slow, as in ~10s.  So, with the k8s API,

- surely the point of the cancelling the context would be to (try to) cancel the asynchronous deletion operation?

- otherwise what is the point of the context at all?

Now here I appreciate that we have the Calico API and not k8s.  But all the same it doesn't seem to make sense.

- If the desired operation has always completed before the Delete call returns, why should we ask the caller to cancel the context, when our API code could just do that itself?

- If the desired operation isn't known to have completed, we shouldn't risk cancelling the context so soon.

I'm probably missing something...

Reply:
The purpose of the context is to allow the IO that the `Delete()` call does to be cancelled (in this case, I'm just using that to impose a timeout, but, in general, you could call `cancel()` from another goroutine). 

The delete returns after doing the write to the datastore.  If there are watchers, they will be asynchronously notified of update (the context doesn't have anything to do with that, though).

The API call can't cancel the context, because it doesn't have the `cancel()` function; in general, a context receiver isn't supposed to cancel the context, because it doesn't know how broad the context is.  For example, you might want to do several IO operations sequentially all within a single timeout.  A classic example is a web server creating a context for each HTTP request with a processing timeout:

- If the sum of all the backend operations is too long, the context times out and all the backend operations get stopped.

- If the user kills the connection then the request machinery can call cancel() to pro-actively stop everything.

Reply:
Thanks for explaining - this all makes sense now.